### PR TITLE
Feature/SemanticID

### PR DESCRIPTION
**Pull request by yzhou-bcom**

### CODE REVIEW

File: interfaces/SolAR3DPointsViewerOpengl.h [hunk 1]
```diff
@@ -173,6 +173,15 @@ class SOLAROPENGL_EXPORT_API SolAR3DPointsViewerOpengl : public org::bcom::xpcf:
     /// @brief points color for the second cloud
     std::vector<unsigned int> m_points2Color = {255, 0, 0};
 
+    /// @brief display points color according to class label 
+    unsigned int m_pointsColorFromClassLabel = 0;
```
Comment (jim-bcom) [inline interfaces/SolAR3DPointsViewerOpengl.h]: Shouldn't this be a boolean instead? If so, maybe rename it to something more bool, like `m_usePointsColorFromClassLabel` ?

File: src/SolAR3DPointsViewerOpengl.cpp [hunk 2]
```diff
@@ -84,6 +86,27 @@ xpcf::XPCFErrorCode SolAR3DPointsViewerOpengl::onConfigured()
     m_resolutionX = m_width;
     m_resolutionY = m_height;
 
+    if (m_pointsColorFromClassLabel>0 && !m_classLabelColorMapPath.empty()) {
```
Comment (jim-bcom) [inline src/SolAR3DPointsViewerOpengl.cpp]: nit: add log message if file path is not filled.
Ex (you can change message/error code)
```
if (m_pointsColorFromClassLabel)
{
  if( m_classLabelColorMapPath.empty() )
  {
    LOG_ERROR("Property classLabelColorMapPath is not defined and is needed when pointsColorFromClassLabel is > 0");
    return xpcf::XPCFErrorCode::_FAIL;
  }
```

File: src/SolAR3DPointsViewerOpengl.cpp [hunk 3]
```diff
@@ -84,6 +86,27 @@ xpcf::XPCFErrorCode SolAR3DPointsViewerOpengl::onConfigured()
...
+        std::ifstream colorFptr;
+        colorFptr.open(m_classLabelColorMapPath);
+        if (!colorFptr.is_open()) {
+            LOG_ERROR("failed to open color map file");
```
Comment (jim-bcom) [inline src/SolAR3DPointsViewerOpengl.cpp]: Add file path in log error message

File: src/SolAR3DPointsViewerOpengl.cpp [hunk 4]
```diff
@@ -329,12 +369,28 @@ void SolAR3DPointsViewerOpengl::OnRender()
         glPointSize(m_pointSize);
         glBegin(GL_POINTS);
         for (unsigned int i = 0; i < m_points.size(); ++i) {
-         if (m_fixedPointsColor)
-            glColor3f(m_pointsColor[0], m_pointsColor[1], m_pointsColor[2]);
-         else
-             glColor3f(m_points[i]->getR(), m_points[i]->getG(), m_points[i]->getB());
-
-         glVertex3f(m_points[i]->getX(), -m_points[i]->getY(), -m_points[i]->getZ());
+            // if color map is provided, display point cloud according to class colors
```
Comment (jim-bcom) [inline src/SolAR3DPointsViewerOpengl.cpp]: Looks similar to code above, is it possible to factorize it in a function/lambda ?

File: src/SolAR3DPointsViewerOpengl.cpp [hunk 5]
```diff
@@ -311,12 +334,29 @@ void SolAR3DPointsViewerOpengl::OnRender()
...
-
-			glVertex3f(m_points2[i]->getX(), -m_points2[i]->getY(), -m_points2[i]->getZ());
+                    // if color map is provided, display point cloud according to class colors
+                    // if cloud point does not have semantic id, display it in white color 
+                    if (m_pointsColorFromClassLabel>0) {
+                        if (m_points2[i]->getSemanticId() < 0) { // no semantic id associated
+                            glColor3f(1.f, 1.f, 1.f); // show in white color 
+                        }
+                        else if (m_points2[i]->getSemanticId() >= static_cast<int>(m_colorMap.size())) {
+                            LOG_ERROR("Cloud point's semantic id exceeds the number of colors");
```
Comment (jim-bcom) [inline src/SolAR3DPointsViewerOpengl.cpp]: Maybe log value of semantic id and number of colors (colormap size) ?

File: src/SolAR3DPointsViewerOpengl.cpp (hunk 2 above)
Comment (yzhou-bcom) [reply src/SolAR3DPointsViewerOpengl.cpp]: done

File: src/SolAR3DPointsViewerOpengl.cpp (hunk 3 above)
Comment (yzhou-bcom) [reply src/SolAR3DPointsViewerOpengl.cpp]: idem

File: src/SolAR3DPointsViewerOpengl.cpp (hunk 4 above)
Comment (yzhou-bcom) [reply src/SolAR3DPointsViewerOpengl.cpp]: lambda functio fnAssignColor is now used

File: src/SolAR3DPointsViewerOpengl.cpp (hunk 5 above)
Comment (yzhou-bcom) [reply src/SolAR3DPointsViewerOpengl.cpp]: done

**Review by jim-bcom — APPROVED** [review]

**Pull request merged**